### PR TITLE
[MDEPLOY-206] Support parallel deploy at end

### PR DESCRIPTION
**Pull request by mattnelson**
https://issues.apache.org/jira/browse/MDEPLOY-206
Support parallel deploy at end.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MPH) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MPH-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MPH-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/deploy/DeployMojo.java [hunk 1]
```diff
@@ -170,12 +183,71 @@ public void execute()
...
+                            return projectDeployerRequest.getProject().getName();
+                        }
+                    };
+
+                    if ( threads > 1 )
+                    {
+                        getLog().info( "Submitting " + projectDeployerRequest.getProject().getName() );
+                        service.submit( callable );
+                    }
+                    else
```
Comment (digulla) [inline src/main/java/org/apache/maven/plugins/deploy/DeployMojo.java]: The `else` could be avoided by using an executor with just a single worker thread.

Comment (mattnelson) [reply src/main/java/org/apache/maven/plugins/deploy/DeployMojo.java]: I did this to maintain passivity by leaving the upload on the current thread.

Comment (digulla) [reply src/main/java/org/apache/maven/plugins/deploy/DeployMojo.java]: You could do that by using a ExecutionService which runs the workers in the same thread like the DirectExecutor from https://google.github.io/guava/releases/24.0-jre/api/docs/index.html?com/google/common/util/concurrent/MoreExecutors.html

Comment (olamy) [reply src/main/java/org/apache/maven/plugins/deploy/DeployMojo.java]: please no dependency on guava....

Comment (digulla) [reply src/main/java/org/apache/maven/plugins/deploy/DeployMojo.java]: I'd create a similar service in Maven's support libraries. Being able to switch freely between parallel and serial execution with the same code is probably useful elsewhere (like running unit tests or reports).

**Comment by slawekjaranowski** [comment]
@mattnelson thanks for your effort on resolving issue.
Last time we have done many code refactor, so implementing was more easier. 

Suppressed by #35

**Comment by jira-importer** [comment]
Resolve #480

**Pull request closed without merging**